runtime/browser/xwalk_render_message_filter.cc

```diff
@@ -32,24 +35,46 @@ bool XWalkRenderMessageFilter::OnMessageReceived(
     IPC_MESSAGE_HANDLER(ViewMsg_OpenLinkExternal, OnOpenLinkExternal)
 #if defined(OS_ANDROID)
     IPC_MESSAGE_HANDLER(XWalkViewHostMsg_SubFrameCreated, OnSubFrameCreated)
+    IPC_MESSAGE_HANDLER(XWalkViewHostMsg_ShouldOverrideUrlLoading,
+                        OnShouldOverrideUrlLoading)
 #endif
     IPC_MESSAGE_UNHANDLED(handled = false)
   IPC_END_MESSAGE_MAP()
 
   return handled;
 }
 
-void XWalkRenderMessageFilter::OnOpenLinkExternal(const GURL& url) {
-  LOG(INFO) << "OpenLinkExternal: " << url.spec();
-  platform_util::OpenExternal(url);
-}
-
 #if defined(OS_ANDROID)
 void XWalkRenderMessageFilter::OnSubFrameCreated(
     int parent_render_frame_id, int child_render_frame_id) {
   XWalkContentsIoThreadClient::SubFrameCreated(
       process_id_, parent_render_frame_id, child_render_frame_id);
 }
+
+void XWalkRenderMessageFilter::OnShouldOverrideUrlLoading(
+    int render_frame_id,
+    const base::string16& url,
+    bool has_user_gesture,
+    bool is_redirect,
+    bool is_main_frame,
+    bool* ignore_navigation) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  *ignore_navigation = false;
+  XWalkContentsClientBridgeBase* client =
+      XWalkContentsClientBridgeBase::FromRenderFrameID(process_id_, render_frame_id);
+  if (client) {
+    *ignore_navigation = client->ShouldOverrideUrlLoading(
+        url, has_user_gesture, is_redirect, is_main_frame);
+  } else {
+    LOG(WARNING) << "Failed to find the associated render view host for url: "
+                 << url;
+  }
+}
 #endif
 
+void XWalkRenderMessageFilter::OnOpenLinkExternal(const GURL& url) {
```

Comment:
I prefer to keep those unchanged, if you want to correct the format, it would be better to submit a separate PR for that.